### PR TITLE
HANS-268: Integrate create care recipient page with serverless component

### PR DESCRIPTION
**Pull request by michal-kras**

### CODE REVIEW

File: management_interface/management_interface/admin.py [hunk 1]
```diff
@@ -23,18 +23,13 @@ class CareRecipientAdmin(admin.ModelAdmin):
         "provider_reference_id",
     )
     list_filter = ("care_provider_location_id",)
+    list_display = [field.name for field in CareRecipient._meta.get_fields()]
```
[the remark this answers is not in the document]
Comment (michal-kras) [reply management_interface/management_interface/admin.py:26]: You mean I can use `list_display = []`? Or what's the `default`?

File: management_interface/management_interface/test_forms.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
+from unittest import mock
```
[the remark this answers is not in the document]
Comment (michal-kras) [reply management_interface/management_interface/test_forms.py:1]: Hmm, I don't understand - What is internal Django tests "mock"?

File: management_interface/management_interface/test_forms.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+    def test_form_displays_necessary_fields(self):
+        form = CareRecipientForm()
+        form_html_lowered = str(form).lower()
+        assert "family name" in form_html_lowered
+        assert "given name" in form_html_lowered
+        assert "nhs number" in form_html_lowered
+        assert "birth" in form_html_lowered
+
+    def test_creating_care_recipient(self):
```
[the remark this answers is not in the document]
Comment (michal-kras) [reply management_interface/management_interface/test_forms.py:28]: I think it would be testing Django Internals (so I didn't find it necessary)

File: management_interface/internal_integrations/management_api/client.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+                        }
+                    ]
+                }
+            ],
+            "name": [
+                {
+                    "use": "official",
+                    "family": patient_family_name,
+                    "given": [
+                        patient_given_name,
```
Comment (henrywrightPSC) [inline management_interface/internal_integrations/management_api/client.py:47]: At this point it should be a list of given names, not just a single given name (e.g. Henry David should become "Henry", "David")

File: management_interface/internal_integrations/management_api/client.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+                                    }
+                                ]
+                            }
+                        }
+                    ]
+                }
+            ],
+            "name": [
+                {
+                    "use": "official",
```
Comment (henrywrightPSC) [inline management_interface/internal_integrations/management_api/client.py:44]: Can we change this to "usual" - it's my mistake in the API specification but just want to keep consistent throughout as we expect hospital and PDS to send us things with "usual".
```suggestion
                    "use": "usual",
```

File: management_interface/management_interface/test_forms.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (grlu75) [reply management_interface/management_interface/test_forms.py:1]: I mean there's a Django testing suite we're using elsewhere. This way we have two different types (styles) of tests. If it's unavoidable to use Mock, then it's OK.

File: management_interface/management_interface/test_forms.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (grlu75) [reply management_interface/management_interface/test_forms.py:28]: OK.

File: management_interface/management_interface/forms.py [hunk 6]
```diff
@@ -19,9 +25,49 @@ class Meta:
 
 
 class CareRecipientForm(forms.ModelForm):
+    given_name = forms.CharField(max_length=64, required=True, label="Given Name")
+    family_name = forms.CharField(max_length=64, required=True, label="Family Name")
+    nhs_number = forms.CharField(max_length=16, required=True, label="NHS Number")
+    birth_date = forms.DateField(required=True, label="Birth Date")
```
[the remark this answers is not in the document]
Comment (grlu75) [reply management_interface/management_interface/forms.py:31]: I'm mostly concerned about the user errors. So definitely hint or label would be great.

File: management_interface/management_interface/admin.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (michal-kras) [reply management_interface/management_interface/admin.py:26]: I don't think it does. Default behaviour is `__str__` of an object, list display displays each field in separate column

File: management_interface/management_interface/forms.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (michal-kras) [reply management_interface/management_interface/forms.py:31]: Good idea

File: management_interface/internal_integrations/management_api/client.py (hunk 4 above)
Comment (michal-kras) [reply management_interface/internal_integrations/management_api/client.py:47]: Hmm, how do you want this to be displayed as an input? One text field, then splitting by space?

File: management_interface/management_interface/forms.py [hunk 7]
```diff
@@ -19,9 +25,49 @@ class Meta:
...
+
+    def save(self, commit: bool = True):
+        self.instance.subscription_id = self.cleaned_data["subscription_id"]
+        self.instance.nhs_number_hash = self.cleaned_data["nhs_number_hash"]
+        return super().save(commit=commit)
+
+    def _create_subscription(self) -> uuid.UUID:
+        try:
+            return ManagementAPIClient().create_subscription(
+                patient_given_name=self.cleaned_data["given_name"],
```
[the remark this answers is not in the document]
Comment (michal-kras) [reply management_interface/management_interface/forms.py:56]: Good point, for leading/trailing whitespaces Django will remove them. For middle spaces let's wait for Henry's opinion in comment below

File: management_interface/internal_integrations/management_api/client.py (hunk 4 above)
Comment (henrywrightPSC) [reply management_interface/internal_integrations/management_api/client.py:47]: @michal-kras can we have one text field and split by spaces - this will match the bulk upload behaviour

File: management_interface/management_interface/forms.py [hunk 8]
```diff
@@ -19,9 +25,46 @@ class Meta:
 
 
 class CareRecipientForm(forms.ModelForm):
+    given_name = forms.CharField(max_length=64, required=True, label="Given Name")
+    family_name = forms.CharField(max_length=64, required=True, label="Family Name")
+    nhs_number = forms.CharField(max_length=16, required=True, label="NHS Number", help_text="e.g. 5991993823")
```
Comment (henrywrightPSC) [inline management_interface/management_interface/forms.py:30]:
```suggestion
    nhs_number = forms.CharField(max_length=12, required=True, label="NHS Number", help_text="e.g. 999 999 9999")
```

This should have max length of 12 to allow for entry without or with spaces. Modified help text since NHS Numbers should always be presented to the user in 3,3,4 format

File: management_interface/management_interface/forms.py [hunk 9]
```diff
@@ -19,9 +25,46 @@ class Meta:
 
 
 class CareRecipientForm(forms.ModelForm):
+    given_name = forms.CharField(max_length=64, required=True, label="Given Name")
```
Comment (henrywrightPSC) [inline management_interface/management_interface/forms.py:28]: Change to reflect needing to enter all given names

```suggestion
    given_name = forms.CharField(max_length=64, required=True, label="Given Names", help_text="enter all given (first and middle) names, separated by spaces, e.g. Lavina Maxine")
```

**Review by henrywrightPSC — CHANGES_REQUESTED** [review]
Made some small suggestions, mainly around ensuring the names are valid in the FHIR Patient we are sending on (through splitting up given names).

File: management_interface/management_interface/admin.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (grlu75) [reply management_interface/management_interface/admin.py:26]: You're right. Looked into wrong place. 
@henrywrightPSC Do we want to show all the columns? The amount of data is huge.

[the remark this answers is not in the document]
Comment (henrywrightPSC) [reply management_interface/management_interface/admin.py:26]: Good Q @grlu75 - could we show only the following?
* `provider_reference_id` (i.e. the one they input themselves)
* `care_provider_location`
* `updated_at`
* `updated_by`

[the remark this answers is not in the document]
Comment (grlu75) [reply management_interface/management_interface/admin.py:26]: Yes, we can filter whatever we want. We also should use __str__ method, as the amount of alphanumerals is overhelming.

[the remark this answers is not in the document]
Comment (grlu75) [reply management_interface/management_interface/admin.py:26]: Probably the best we can do is to split this to another small task and do it later.

File: management_interface/management_interface/forms.py [hunk 10]
```diff
@@ -19,9 +25,46 @@ class Meta:
...
+            return ManagementAPIClient().create_subscription(
+                patient_given_name=self.cleaned_data["given_name"],
+                patient_family_name=self.cleaned_data["family_name"],
+                nhs_number=self.cleaned_data["nhs_number"],
+                birth_date=self.cleaned_data["birth_date"],
+            )
+        except ManagementAPIClientError as ex:
+            raise ValidationError(str(ex))
+
+    def _generate_nhs_number_hash(self) -> str:
```
Comment (regularfry) [inline management_interface/management_interface/forms.py:61]: Moving this here breaks the bulk upload, right?  What are we doing about that?  [This](https://github.com/nhsx/il-hans-management-interface/blob/main/management_interface/management_interface/admin.py#L84) relies on the model itself calculating the hash.

**Review by regularfry — CHANGES_REQUESTED** [review]
Same as #13 - tests fail for me.  But also I'm not sure moving the nhs number hashing out of the model is the right thing to do.

File: management_interface/internal_integrations/management_api/client.py (hunk 4 above)
Comment (michal-kras) [reply management_interface/internal_integrations/management_api/client.py]: @regularfry CSV upload hasn't been merged when I was developing this, I will rebase and see what can be done.

File: management_interface/management_interface/forms.py (hunk 10 above)
Comment (michal-kras) [reply management_interface/management_interface/forms.py:75]: Same: https://github.com/nhsx/il-hans-management-interface/pull/11#discussion_r1156214351

**Comment by michal-kras** [comment]
> Same as #13 - tests fail for me. But also I'm not sure moving the nhs number hashing out of the model is the right thing to do.

imo model should reflect what's stored in database. Since we do not store plain-text nhs numbers, the model shouldn't do the job of controller.

**Comment by michal-kras** [comment]
> Same as #13 - tests fail for me. But also I'm not sure moving the nhs number hashing out of the model is the right thing to do.

Exact command you provided works for me, can you retry with clean docker rebuilt project?

File: management_interface/internal_integrations/management_api/client.py (hunk 4 above)
Comment (regularfry) [reply management_interface/internal_integrations/management_api/client.py]: Yep, no worries - perils of not integrating often enough.  It happens.

**Comment by regularfry** [comment]
> > Same as #13 - tests fail for me. But also I'm not sure moving the nhs number hashing out of the model is the right thing to do.
> 
> imo model should reflect what's stored in database. Since we do not store plain-text nhs numbers, the model shouldn't do the job of controller (in our case form).

I don't disagree, but if we're going to say "the model should only concern itself with the database storage" then we should also move subscription API responsibility out of it for consistency.  From a conceptual viewpoint either the Django model should take strong responsibility for all the business logic (including storage) or we should identify a specific alternative place that logic should live, and leave the Django model to only handle the database.

This is what I meant by cutting against the grain: I get the feeling that the more logic we move out of the Django model, the more we're going to be tripping up the admin interface. Happy to be proved wrong if we've got a good pattern for doing it, but I'm worried about creating work for ourselves by having to work around too much of Django to make it worthwhile.

**Comment by michal-kras** [comment]
@henrywrightPSC @regularfry 

> Django is reasonable strong on this and says that all relevant domain logic belongs with the models - [see the design pattern here on active records](https://docs.djangoproject.com/en/4.2/misc/design-philosophies/)

Additional field (`plain-texted` nhs_number) is not domain logic though. Going by that, we should also couple the model with dob, given_name, family_name fields which doesn't make any sense (imo).

If we were storing plain-text nhs_number, then the logic of implementing a function masking it in the output should be a part of a model.

> I don't disagree, but if we're going to say "the model should only concern itself with the database storage" then we should also move subscription API responsibility out of it for consistency.

Completely agree on that point, and it's not a part of model logic. I did use a `Form` as a controller, perhaps it should be moved away to separate module/class? (I don't think it's necessary, as `Form` in Django quite nicely fills the controller role).

**Comment by henrywrightPSC** [comment]
> @henrywrightPSC @regularfry
> 
> > Django is reasonable strong on this and says that all relevant domain logic belongs with the models - [see the design pattern here on active records](https://docs.djangoproject.com/en/4.2/misc/design-philosophies/)
> 
> Additional field (`plain-texted` nhs_number) is not domain logic though. Going by that, we should also couple the model with dob, given_name, family_name fields which doesn't make any sense (imo).
> 
> If we were storing plain-text nhs_number, then the logic of implementing a function masking it in the output should be a part of a model.
> 
> > I don't disagree, but if we're going to say "the model should only concern itself with the database storage" then we should also move subscription API responsibility out of it for consistency.
> 
> Completely agree on that point, and it's not a part of model logic. I did use a `Form` as a controller, perhaps it should be moved away to separate module/class? (I don't think it's necessary, as `Form` in Django quite nicely fills the controller role).

I'm convinced @michal-kras!

**Review by henrywrightPSC — APPROVED** [review]

**Comment by regularfry** [comment]
> @henrywrightPSC @regularfry
> 
> > Django is reasonable strong on this and says that all relevant domain logic belongs with the models - [see the design pattern here on active records](https://docs.djangoproject.com/en/4.2/misc/design-philosophies/)
> 
> Additional field (`plain-texted` nhs_number) is not domain logic though. Going by that, we should also couple the model with dob, given_name, family_name fields which doesn't make any sense (imo).
> 
> If we were storing plain-text nhs_number, then the logic of implementing a function masking it in the output should be a part of a model.

This is exactly it: we *don't* want a field, but Django only wants us to pass data into the model's scope via fields.  All we really want is a `def set_nhs_number` method but hooking that into Django is painful.

I think using the `Form` as the controller works.

**Comment by regularfry** [comment]
Nearly there...

```
 $ docker-compose exec web python manage.py test
Found 25 test(s).
Creating test database for alias 'default'...
System check identified no issues (0 silenced).
.................EE......
======================================================================
ERROR: test_admin_upload_csv_file_successfully (management_interface.test_views.AdminCareProviderLocationTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/app/management_interface/test_views.py", line 133, in test_admin_upload_csv_file_successfully
    response = self._get_upload_file_response(csv_file)
  File "/app/management_interface/test_views.py", line 68, in _get_upload_file_response
    response = self.client.post(
  File "/usr/local/lib/python3.10/site-packages/django/test/client.py", line 852, in post
    response = super().post(
  File "/usr/local/lib/python3.10/site-packages/django/test/client.py", line 441, in post
    return self.generic(
  File "/usr/local/lib/python3.10/site-packages/django/test/client.py", line 541, in generic
    return self.request(**r)
  File "/usr/local/lib/python3.10/site-packages/django/test/client.py", line 810, in request
    self.check_exception(response)
  File "/usr/local/lib/python3.10/site-packages/django/test/client.py", line 663, in check_exception
    raise exc_value
  File "/usr/local/lib/python3.10/site-packages/django/core/handlers/exception.py", line 56, in inner
    response = get_response(request)
  File "/usr/local/lib/python3.10/site-packages/django/core/handlers/base.py", line 197, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/usr/local/lib/python3.10/site-packages/django/utils/decorators.py", line 134, in _wrapped_view
    response = view_func(request, *args, **kwargs)
  File "/usr/local/lib/python3.10/site-packages/django/views/decorators/cache.py", line 62, in _wrapped_view_func
    response = view_func(request, *args, **kwargs)
  File "/usr/local/lib/python3.10/site-packages/django/contrib/admin/sites.py", line 242, in inner
    return view(request, *args, **kwargs)
  File "/app/management_interface/admin.py", line 124, in import_care_recipients
    created_care_recipients, errors = self._bulk_create_care_recipients(
  File "/app/management_interface/admin.py", line 183, in _bulk_create_care_recipients
    if form.errors:
  File "/usr/local/lib/python3.10/site-packages/django/forms/forms.py", line 200, in errors
    self.full_clean()
  File "/usr/local/lib/python3.10/site-packages/django/forms/forms.py", line 438, in full_clean
    self._clean_form()
  File "/usr/local/lib/python3.10/site-packages/django/forms/forms.py", line 459, in _clean_form
    cleaned_data = self.clean()
  File "/app/management_interface/forms.py", line 56, in clean
    self.cleaned_data["subscription_id"] = self._create_subscription()
  File "/app/management_interface/forms.py", line 67, in _create_subscription
    return ManagementAPIClient().create_subscription(
  File "/app/internal_integrations/management_api/client.py", line 17, in __init__
    self.base_url = base_url or get_management_api_settings().base_url
  File "/app/internal_integrations/management_api/settings.py", line 15, in get_management_api_settings
    return ManagementAPISettings()
  File "/usr/local/lib/python3.10/site-packages/pydantic/env_settings.py", line 39, in __init__
    super().__init__(
  File "/usr/local/lib/python3.10/site-packages/pydantic/main.py", line 341, in __init__
    raise validation_error
pydantic.error_wrappers.ValidationError: 1 validation error for ManagementAPISettings
base_url
  field required (type=value_error.missing)

======================================================================
ERROR: test_admin_upload_csv_file_with_broken_row (management_interface.test_views.AdminCareProviderLocationTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/app/management_interface/test_views.py", line 142, in test_admin_upload_csv_file_with_broken_row
    response = self._get_upload_file_response(csv_file)
  File "/app/management_interface/test_views.py", line 68, in _get_upload_file_response
    response = self.client.post(
  File "/usr/local/lib/python3.10/site-packages/django/test/client.py", line 852, in post
    response = super().post(
  File "/usr/local/lib/python3.10/site-packages/django/test/client.py", line 441, in post
    return self.generic(
  File "/usr/local/lib/python3.10/site-packages/django/test/client.py", line 541, in generic
    return self.request(**r)
  File "/usr/local/lib/python3.10/site-packages/django/test/client.py", line 810, in request
    self.check_exception(response)
  File "/usr/local/lib/python3.10/site-packages/django/test/client.py", line 663, in check_exception
    raise exc_value
  File "/usr/local/lib/python3.10/site-packages/django/core/handlers/exception.py", line 56, in inner
    response = get_response(request)
  File "/usr/local/lib/python3.10/site-packages/django/core/handlers/base.py", line 197, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/usr/local/lib/python3.10/site-packages/django/utils/decorators.py", line 134, in _wrapped_view
    response = view_func(request, *args, **kwargs)
  File "/usr/local/lib/python3.10/site-packages/django/views/decorators/cache.py", line 62, in _wrapped_view_func
    response = view_func(request, *args, **kwargs)
  File "/usr/local/lib/python3.10/site-packages/django/contrib/admin/sites.py", line 242, in inner
    return view(request, *args, **kwargs)
  File "/app/management_interface/admin.py", line 124, in import_care_recipients
    created_care_recipients, errors = self._bulk_create_care_recipients(
  File "/app/management_interface/admin.py", line 183, in _bulk_create_care_recipients
    if form.errors:
  File "/usr/local/lib/python3.10/site-packages/django/forms/forms.py", line 200, in errors
    self.full_clean()
  File "/usr/local/lib/python3.10/site-packages/django/forms/forms.py", line 438, in full_clean
    self._clean_form()
  File "/usr/local/lib/python3.10/site-packages/django/forms/forms.py", line 459, in _clean_form
    cleaned_data = self.clean()
  File "/app/management_interface/forms.py", line 56, in clean
    self.cleaned_data["subscription_id"] = self._create_subscription()
  File "/app/management_interface/forms.py", line 67, in _create_subscription
    return ManagementAPIClient().create_subscription(
  File "/app/internal_integrations/management_api/client.py", line 17, in __init__
    self.base_url = base_url or get_management_api_settings().base_url
  File "/app/internal_integrations/management_api/settings.py", line 15, in get_management_api_settings
    return ManagementAPISettings()
  File "/usr/local/lib/python3.10/site-packages/pydantic/env_settings.py", line 39, in __init__
    super().__init__(
  File "/usr/local/lib/python3.10/site-packages/pydantic/main.py", line 341, in __init__
    raise validation_error
pydantic.error_wrappers.ValidationError: 1 validation error for ManagementAPISettings
base_url
  field required (type=value_error.missing)

----------------------------------------------------------------------
Ran 25 tests in 1.603s

FAILED (errors=2)
Destroying test database for alias 'default'...
```

Missed a commit?

**Comment by michal-kras** [comment]
> Nearly there...
> 
> ```
>  $ docker-compose exec web python manage.py test
> Found 25 test(s).
> Creating test database for alias 'default'...
> System check identified no issues (0 silenced).
> .................EE......
> ======================================================================
> ERROR: test_admin_upload_csv_file_successfully (management_interface.test_views.AdminCareProviderLocationTests)
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/app/management_interface/test_views.py", line 133, in test_admin_upload_csv_file_successfully
>     response = self._get_upload_file_response(csv_file)
>   File "/app/management_interface/test_views.py", line 68, in _get_upload_file_response
>     response = self.client.post(
>   File "/usr/local/lib/python3.10/site-packages/django/test/client.py", line 852, in post
>     response = super().post(
>   File "/usr/local/lib/python3.10/site-packages/django/test/client.py", line 441, in post
>     return self.generic(
>   File "/usr/local/lib/python3.10/site-packages/django/test/client.py", line 541, in generic
>     return self.request(**r)
>   File "/usr/local/lib/python3.10/site-packages/django/test/client.py", line 810, in request
>     self.check_exception(response)
>   File "/usr/local/lib/python3.10/site-packages/django/test/client.py", line 663, in check_exception
>     raise exc_value
>   File "/usr/local/lib/python3.10/site-packages/django/core/handlers/exception.py", line 56, in inner
>     response = get_response(request)
>   File "/usr/local/lib/python3.10/site-packages/django/core/handlers/base.py", line 197, in _get_response
>     response = wrapped_callback(request, *callback_args, **callback_kwargs)
>   File "/usr/local/lib/python3.10/site-packages/django/utils/decorators.py", line 134, in _wrapped_view
>     response = view_func(request, *args, **kwargs)
>   File "/usr/local/lib/python3.10/site-packages/django/views/decorators/cache.py", line 62, in _wrapped_view_func
>     response = view_func(request, *args, **kwargs)
>   File "/usr/local/lib/python3.10/site-packages/django/contrib/admin/sites.py", line 242, in inner
>     return view(request, *args, **kwargs)
>   File "/app/management_interface/admin.py", line 124, in import_care_recipients
>     created_care_recipients, errors = self._bulk_create_care_recipients(
>   File "/app/management_interface/admin.py", line 183, in _bulk_create_care_recipients
>     if form.errors:
>   File "/usr/local/lib/python3.10/site-packages/django/forms/forms.py", line 200, in errors
>     self.full_clean()
>   File "/usr/local/lib/python3.10/site-packages/django/forms/forms.py", line 438, in full_clean
>     self._clean_form()
>   File "/usr/local/lib/python3.10/site-packages/django/forms/forms.py", line 459, in _clean_form
>     cleaned_data = self.clean()
>   File "/app/management_interface/forms.py", line 56, in clean
>     self.cleaned_data["subscription_id"] = self._create_subscription()
>   File "/app/management_interface/forms.py", line 67, in _create_subscription
>     return ManagementAPIClient().create_subscription(
>   File "/app/internal_integrations/management_api/client.py", line 17, in __init__
>     self.base_url = base_url or get_management_api_settings().base_url
>   File "/app/internal_integrations/management_api/settings.py", line 15, in get_management_api_settings
>     return ManagementAPISettings()
>   File "/usr/local/lib/python3.10/site-packages/pydantic/env_settings.py", line 39, in __init__
>     super().__init__(
>   File "/usr/local/lib/python3.10/site-packages/pydantic/main.py", line 341, in __init__
>     raise validation_error
> pydantic.error_wrappers.ValidationError: 1 validation error for ManagementAPISettings
> base_url
>   field required (type=value_error.missing)
> 
> ======================================================================
> ERROR: test_admin_upload_csv_file_with_broken_row (management_interface.test_views.AdminCareProviderLocationTests)
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/app/management_interface/test_views.py", line 142, in test_admin_upload_csv_file_with_broken_row
>     response = self._get_upload_file_response(csv_file)
>   File "/app/management_interface/test_views.py", line 68, in _get_upload_file_response
>     response = self.client.post(
>   File "/usr/local/lib/python3.10/site-packages/django/test/client.py", line 852, in post
>     response = super().post(
>   File "/usr/local/lib/python3.10/site-packages/django/test/client.py", line 441, in post
>     return self.generic(
>   File "/usr/local/lib/python3.10/site-packages/django/test/client.py", line 541, in generic
>     return self.request(**r)
>   File "/usr/local/lib/python3.10/site-packages/django/test/client.py", line 810, in request
>     self.check_exception(response)
>   File "/usr/local/lib/python3.10/site-packages/django/test/client.py", line 663, in check_exception
>     raise exc_value
>   File "/usr/local/lib/python3.10/site-packages/django/core/handlers/exception.py", line 56, in inner
>     response = get_response(request)
>   File "/usr/local/lib/python3.10/site-packages/django/core/handlers/base.py", line 197, in _get_response
>     response = wrapped_callback(request, *callback_args, **callback_kwargs)
>   File "/usr/local/lib/python3.10/site-packages/django/utils/decorators.py", line 134, in _wrapped_view
>     response = view_func(request, *args, **kwargs)
>   File "/usr/local/lib/python3.10/site-packages/django/views/decorators/cache.py", line 62, in _wrapped_view_func
>     response = view_func(request, *args, **kwargs)
>   File "/usr/local/lib/python3.10/site-packages/django/contrib/admin/sites.py", line 242, in inner
>     return view(request, *args, **kwargs)
>   File "/app/management_interface/admin.py", line 124, in import_care_recipients
>     created_care_recipients, errors = self._bulk_create_care_recipients(
>   File "/app/management_interface/admin.py", line 183, in _bulk_create_care_recipients
>     if form.errors:
>   File "/usr/local/lib/python3.10/site-packages/django/forms/forms.py", line 200, in errors
>     self.full_clean()
>   File "/usr/local/lib/python3.10/site-packages/django/forms/forms.py", line 438, in full_clean
>     self._clean_form()
>   File "/usr/local/lib/python3.10/site-packages/django/forms/forms.py", line 459, in _clean_form
>     cleaned_data = self.clean()
>   File "/app/management_interface/forms.py", line 56, in clean
>     self.cleaned_data["subscription_id"] = self._create_subscription()
>   File "/app/management_interface/forms.py", line 67, in _create_subscription
>     return ManagementAPIClient().create_subscription(
>   File "/app/internal_integrations/management_api/client.py", line 17, in __init__
>     self.base_url = base_url or get_management_api_settings().base_url
>   File "/app/internal_integrations/management_api/settings.py", line 15, in get_management_api_settings
>     return ManagementAPISettings()
>   File "/usr/local/lib/python3.10/site-packages/pydantic/env_settings.py", line 39, in __init__
>     super().__init__(
>   File "/usr/local/lib/python3.10/site-packages/pydantic/main.py", line 341, in __init__
>     raise validation_error
> pydantic.error_wrappers.ValidationError: 1 validation error for ManagementAPISettings
> base_url
>   field required (type=value_error.missing)
> 
> ----------------------------------------------------------------------
> Ran 25 tests in 1.603s
> 
> FAILED (errors=2)
> Destroying test database for alias 'default'...
> ```
> 
> Missed a commit?

Ok, that's a bit of pandora box we're opening here.

TLDR: add env_variable: MANAGEMENT_API_BASE_URL=https://fowvgbazdl.execute-api.eu-west-2.amazonaws.com/Prod/

They should pass, but clearly they are not unit tests anymore:

1. We can live with it
2. We can move them to separate package like: `tests/integration`
3. We can refactor them to mock out network dependencies

WDYT @regularfry ?

**Comment by regularfry** [comment]
In this case I think I'd want a mock of the `ManagementAPIClient` class rather than mocking out the network call.  Don't mock what you don't own, and all that. It'll end up with a better structure.

I think we *also* want an integration test, mind, but the more important thing to me is getting the network call out of the unit test run.

**Comment by michal-kras** [comment]
> In this case I think I'd want a mock of the `ManagementAPIClient` class rather than mocking out the network call. Don't mock what you don't own, and all that. It'll end up with a better structure.
> 
> I think we _also_ want an integration test, mind, but the more important thing to me is getting the network call out of the unit test run.

This sounds good to me, will do.
